src/MatrixTwitter.js

```diff
@@ -0,0 +1,987 @@
+var Twitter = require('twitter');
+var Request = require('request');
+var fs = require('fs');
+var log = require('npmlog');
+var Buffer = require('buffer').Buffer;
+var HTMLDecoder = new require('html-entities').AllHtmlEntities;
+var MatrixRoom = require("matrix-appservice-bridge").MatrixRoom;
+var RemoteRoom = require("matrix-appservice-bridge").RemoteRoom;
+
+var ProcessedTweetList = require("./ProcessedTweetList.js");
+var util = require("./util.js");
+
+const TWITTER_PROFILE_INTERVAL_MS   = 300000;
+const TWITTER_CLIENT_INTERVAL_MS    = 60000;
+const TWITTER_MSG_QUEUE_INTERVAL_MS = 1500;
+const TWEET_REPLY_MAX_DEPTH = 3;
+
+
+/**
+ * MatrixTwitter - This class handles the connections between the Twitter API
+ * and the bridge.
+ * @class
+ * @param  {matrix-appservice-bridge.Bridge} bridge
+ * @param  config. Bridge configuration (currently only make use of the auth
+*  information)
+ * @param  config.app_auth.consumer_key Twitter consumer key
+ * @param  config.app_auth.consumer_secret Twitter consumer secret
+ * @param  {TwitterDB}       storage
+ */
+var MatrixTwitter = function (bridge, config, storage) {
+  this.app_auth = config.app_auth;
+  this.app_twitter = null;
+  this.tclients = {};
+  this.storage = storage;
+
+  this.timeline_list = [];
+  this.timeline_queue = [];
+  this.timeline_period = 3050; //Twitter allows 300 calls per 15 minute (We add 50 milliseconds for a little safety).
+  this.timeline_intervalID = null;
+
+  this.hashtag_period = 2050; //Twitter allows 450 calls per 15 minute (We add 50 milliseconds for a little safety).
+  this.hashtag_intervalID = null;
+  this.hashtag_list = [];
+  this.hashtag_queue = [];
+
+  this.user_streams = {};
+
+  this.processed_tweets = new ProcessedTweetList(256,32);  //This will contain all the tweet IDs of things we don't want to repeat.
+  this.sent_dms = new ProcessedTweetList(1,1);  //This will contain the body of the DM posted to a room to avoid reposting it.
+  this.msg_queue = [];
+  this.msg_queue_intervalID = null;
+  this._bridge = bridge;
+  this.tweet_event_cache = {};
+};
+
+
+/**
+ * MatrixTwitter.prototype.start - Starts the timers for polling the API and
+ * authenticates the application with Twitter.
+ * @function
+ * @return {Promise}  A promise that returns when authentication succeeds.
+ */
+MatrixTwitter.prototype.start = function(){
+  return this._get_bearer_token().then((token) => {
+    this.app_auth.bearer_token = token;
+    log.info('Twitter','Retrieved token');
+    this.app_twitter = new Twitter(this.app_auth);
+    this.msg_queue_intervalID = setInterval(() => {this._process_head_of_msg_queue();}, TWITTER_MSG_QUEUE_INTERVAL_MS);
```

Comment:
Typically if you're going to have a public API which starts a polling routine, you should guard said public API to protect against dumb callers who call the same thing twice. Up to you if you want to ignore subsequent `start()` requests, `log.warn` them or `throw`, but I insist you do *something* because trust me, they are really subtle annoying things to debug should it happen.